### PR TITLE
S3: close DownloadInstanceWriter instream

### LINKED ISSUE #7910: S3: DownloadInstanceWriter may never close InputStream
opening a separate issue for https://github.com/IQSS/dataverse/issues/7803#issuecomment-847344123

@qqmyers suggests the use of `try(InputStream instream = storageIO.getInputStream())`

PR forthcoming.

### PR DESCRIPTION
**Pull request by donsizemore**
**What this PR does / why we need it**: see https://github.com/IQSS/dataverse/issues/7803#issuecomment-847344123

**Which issue(s) this PR closes**:

Closes #7910 

**Special notes for your reviewer**: this commit is best viewed with `&w=1`: https://github.com/IQSS/dataverse/compare/develop...OdumInstitute:7910_S3_close_inputstream?diff=split&w=1

**Suggestions on how to test this**: follow scenario described in https://github.com/IQSS/dataverse/issues/7803#issuecomment-847344123

**Does this PR introduce a user interface change? If mockups are available, please link/include them here**: no

**Is there a release notes update needed for this change?**: no

**Additional documentation**: no

### CODE REVIEW

File: src/main/java/edu/harvard/iq/dataverse/api/DownloadInstanceWriter.java [hunk 1]
```diff
@@ -385,91 +385,92 @@ public void writeTo(DownloadInstance di, Class<?> clazz, Type type, Annotation[]
...
-                    boolean useChunkedTransfer = false;
-                    //if ((contentSize = getFileSize(di, storageIO.getVarHeader())) > 0) {
-                    if ((contentSize = getContentSize(storageIO)) > 0) {
-                        logger.fine("Content size (retrieved from the AccessObject): " + contentSize);
-                        httpHeaders.add("Content-Length", contentSize);
-                    } else {
-                        //httpHeaders.add("Transfer-encoding", "chunked");
-                        //useChunkedTransfer = true;
-                    }
+                try (InputStream instream = storageIO.getInputStream()) {
```
Comment (qqmyers) [inline src/main/java/edu/harvard/iq/dataverse/api/DownloadInstanceWriter.java:388]: Looks good to me. try-with-resources is the preferred way to assure streams are closed these days and that's what the PR does.

File: src/main/java/edu/harvard/iq/dataverse/api/DownloadInstanceWriter.java [hunk 2]
```diff
@@ -385,91 +385,92 @@ public void writeTo(DownloadInstance di, Class<?> clazz, Type type, Annotation[]
...
+                            logger.fine("not writing guestbook response");
                         }
-                    } else {
-                        logger.fine("not writing guestbook response");
-                    }
 
-                    instream.close();
-                    outstream.close();
-                    return;
+                        instream.close();
```
Comment (qqmyers) [inline src/main/java/edu/harvard/iq/dataverse/api/DownloadInstanceWriter.java:470]: I guess this line is now redundant and could be removed.

**Review by qqmyers — CHANGES_REQUESTED** [review]
Don - looks good - can you make it a three-line change instead (see comment)? With that, I think this is good to go.

**Review by qqmyers — APPROVED** [review]
The test @pdurbin saw fail appears to have succeeded in the new run. I can't see any reason the original commit would have caused a problem or why the second would fix it, so I'm guessing the test issue is some intermittent/unrelated issue.

**Review by scolapasta — APPROVED** [review]

**Pull request merged**